mitodl/mitxonline
# Adds program record functionality

#### Pre-Flight checklist


#### What are the relevant tickets?

Fixes #1208 

#### What's this PR do?

Adds learner program records to the system:
- The learner can access their record by accessing the View Program Record link in the program drawer for the program.
- The record shows basic info about the learner and the program, and lists out the courses in the program, with the learner's status in each. 
   - The record has a tag for Partially Completed or Completed depending on their completion of the courses in the program.
   - Each course is tagged with Not Enrolled or Certificate Earned tags appropriately.
   - Each course lists the highest grade the learner achieved in the course, if there is one.
   - Each course is tagged with either Core or Elective, depending on where it sits in the requirements tree.
- The learner can activate record sharing for the record.
   - Upon activation, an anonymous share record is created. This uses a UUID to identify the particular learner record.
   - The learner can activate the Share button to retrieve a link to the anonymous share, which can then be copied and sent to other people at the learner's choosing.
   - The Share dialog has a Copy Link button which will copy the link text to the clipboard.
   - The learner can activate the Send Learner Record button, which allows the learner to send their record directly to a partner school configured within the system. 
      - Partner school shares have their own identifying UUID and exist separately from the anonymous share.
      - Partner schools are alerted to new shared learner records via email.
   - The learner can revoke sharing.
      - Revoking sharing invalidates the active anonymous record share UUID.
      - Revoking sharing _does not_ invalidate any partner records that have been shared. The records can be invalidated or deleted on the admin side, though.
      - Re-sharing the record generates a _new_ UUID for the anonymous share - _old sharing links will never work again_. 
 - Anonymous users can activate a shared link (either from a partner school email or from the learner directly).
    - Anonymous viewing displays the same information as is displayed to the learner.
    - Anonymous viewing does not provide sharing controls for the record. 
- Record sharing is hidden behind a feature flag.
   - The feature flag is `ENABLE_LEARNER_RECORDS`.
   - Additionally, the program UI needs to be turned on - the link to activate is in code that is hidden if the program UI is off.
   - As with the program UI, the learner record code can be enabled by appending `enable_lr` to the URL in some way. (The code that checks for these things checks for the code as part of the GET parameters - if you need to enable both learner records and the program UI via the URL, appending something like `?enable_programs&enable_lr` should work.) 
    
#### How should this be manually tested?

1. Ensure you have a learner account enrolled in a program.
2. Generate grades and certificates for the learner.
3. Using Django Admin, also create some partner school records. 
4. View the learner's dashboard and open the program drawer. You should see the View Program Record link.
5. Activate the link. The Program Record should open in a new tab.
6. Check to see that the courses are displayed properly, the learner's info is correct, and the tags are also correct. 
7. Enable sharing by clicking Allow Access. Once this is done, the buttons at the top should be Send Learner Record, Share, and Revoke Access.
8. Activate the Share button. You should see a link to the learner record, and you should be able to copy it using the button The button should indicate that the copy was successful.
9. In an incognito window or separate browser, navigate to the link that was copied. You should see the learner record without any sharing controls. Save the URL for this for later testing.
10. Activate the Send Learner Record button. You should be presented with a dialog to share the learner record to the schools configured in step 3. 
11. Select a school and send the record. You should receive an email with a link in it. 
12. Activate the link in a separate browser or incognito window. You should see the program record minus any sharing controls.
13. Revoke sharing by clicking Revoke Sharing. You should get a notification that the revocation was successful and the buttons should change back to the single Allow Sharing button.
14. Try the share link from step 8 again. It should fail.
15. Try the share link from the email generated in step 11 again. It should succeed. 
16. Re-activate sharing. Then, click Share. The link should be different. 
17. Reload the program record. The record should render with the full sharing controls enabled (rather than just Allow Sharing). 

#### Screenshots (if appropriate)

Initial record screen: 
![image](https://user-images.githubusercontent.com/945611/204615181-049ae601-a995-4646-8f25-14d13da8d754.png)

After allowing sharing:
![image](https://user-images.githubusercontent.com/945611/204615374-2c6d0ba5-e670-4572-8e55-4f3c263f93a8.png)

Dialog - Send Learner Record:
![image](https://user-images.githubusercontent.com/945611/204615426-74082385-410c-45d7-9951-b9b9aaa18a62.png)

Dialog - Share:
![image](https://user-images.githubusercontent.com/945611/204615473-21eb841c-3f8a-41b6-b871-cd3ac3c96dae.png)

Anonymous View:
![image](https://user-images.githubusercontent.com/945611/204615607-f47c0aeb-6f40-4d66-a286-570ea32c4ada.png)
